### PR TITLE
Codegen: python_torch_functions only include relevant operators

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69840
* #70021
* __->__ #68693

Generation of python bindings for native functions is split over 8
different files. One for each namespace, with the torch namespace
split into 3 shards, and methods in their own file as well. This
change ensures that editing any single (non-method) operator only
causes one of these files to be rebuilt.

Differential Revision: [D32596270](https://our.internmc.facebook.com/intern/diff/D32596270)

### CODE REVIEW

**Review by albanD — APPROVED** [review]
SGTM

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

**Comment by albanD** [comment]
@peterbell10 isn't this breaking the non per-op build? Can you try locally?
In particular, the ops/tensor.h and ops/from_blob.h tensors need to be handled in a specific way?
Do we want to make sure the OSS build works without per-op? Or I should just fix the internal build directly?

**Comment by peterbell10** [comment]
> @peterbell10 isn't this breaking the non per-op build? Can you try locally?

Works for me locally

> In particular, the `ops/tensor.h` and `ops/from_blob.h` tensors need to be handled in a specific way?

They are just regular files and not code generated, so no. You might have to add a new glob or something to the buck build so they get picked up.

> Do we want to make sure the OSS build works without per-op? Or I should just fix the internal build directly?

I would prefer the OSS build works with per operator headers. I've got a PR where I'm trying to add it to the CI. 

What errors are you seeing internally?

**Comment by albanD** [comment]
> What errors are you seeing internally?

Only missing includes.
I though they would be missing from generated outputs. But they are indeed plain files. Let me add them to the buck specs and test again.
Thanks for the quick answer!

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

File: torch/csrc/api/include/torch/types.h [hunk 1]
```diff
@@ -7,6 +7,11 @@
 #include <torch/csrc/autograd/generated/variable_factories.h>
 #include <torch/csrc/autograd/variable.h>
 
+// TODO: These don't really belong here but torchvision builds in CI need them
+// Remove once the torchvision version being compiled in CI is updated
+#include <torch/library.h>
```
Comment (albanD) [inline torch/csrc/api/include/torch/types.h:12]: This include causes problems for other builds. In particular, it prevents including torch.h direclty from what I can see:

```
In file included from original_header.h:7:
In file included from torch/csrc/api/include/torch/torch.h:3:
In file included from torch/csrc/api/include/torch/all.h:8:
In file included from torch/csrc/api/include/torch/data.h:3:
In file included from torch/csrc/api/include/torch/data/dataloader.h:3:
In file included from torch/csrc/api/include/torch/data/dataloader/stateful.h:3:
In file included from torch/csrc/api/include/torch/data/dataloader/base.h:3:
In file included from torch/csrc/api/include/torch/data/dataloader_options.h:4:
In file included from torch/csrc/api/include/torch/types.h:12:
xplat/caffe2/torch/library.h:397:39: error: no member named 'parseSchema' in namespace 'torch::jit'
  c10::FunctionSchema s = torch::jit::parseSchema(str);
                          ~~~~~~~~~~~~^
# More errors like that
```

Removing this include does allow the build to go smoothly.

Comment (peterbell10) [reply torch/csrc/api/include/torch/types.h:12]: Looks like the include of `function_schema_parser.h` was guarded but the uses of it weren't. The old code probably included it transitively via `variable.h` or something like that.

Comment (albanD) [reply torch/csrc/api/include/torch/types.h:12]: Thanks for finding this out!

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596270).

**Comment by peterbell10** [comment]
Regarding the recent CI failures, I wasn't able to reproduce it until after doing `conda update --all` and it still fails in branches that don't include this PR, so I'm confident it's an unrelated dependency issue.

**Pull request closed without merging**